### ISSUE TITLE
[Suggestions] Improvements on the docker image

### ISSUE THREAD

**Issue by KurtzPT** [MrBrax/TwitchAutomator#89]
Nice things to be improved (but not totally necessary):

- Allow us to set up the UID/GID/UMASK in the container to apply on files that are written inside the volumes of the docker that TwitchAutomator uses.
 
At the moment the permissions are a mess, I cannot delete for example vods directly from the folders (or any file), because they are set to this user "NOBODY" that is unknown outside of docker. (Unraid uses another user "NOBODY" with a different UID). Currently I have to setup a cron job in root to chmod/chown these volume folders/files every X minutes.

Usually we can pass these values using environment variables, you can take a look at most linuxserver docker images here in github, they always have this so we don't have issues with permissions.

- Same thing can be said for the volumes mappings, I always have to edit out the docker-compose file myself, instead of passing the values to override.

- Since you changed the docker to use PHP8, there are some settings in Dockerfile and Composer that are still using PHP7.

**Comment by MrBrax** [comment]
good idea, i can check out those permission things

i think i'll keep it php7 compatible as long as my host still uses it, so no forcing yet

**Comment by MrBrax** [comment]
https://github.com/TrafeX/docker-php-nginx/issues/43

I think permissions could be problematic, could maybe work if you built the containers manually every time

**Comment by KurtzPT** [comment]
Instead of changing the user who runs the php stuff, maybe we can set the group user of each volume by defining it via env variables? And making it read/write permissions on every single file inside these volumes for that specific group ?

This way basically does the objective which is being able to add/edit/remove files from the volumes myself without using a root user.

**Comment by MrBrax** [comment]
is it not possible to run docker as a different user?

**Comment by KurtzPT** [comment]
> is it not possible to run docker as a different user?

You mean the docker process that runs all containers? If that's the case I have no idea, but I would rather not mess up with the unraid configuration for it.

**Comment by MrBrax** [comment]
yeah something like that, seems like the only other solution for in-container permissions is to set it up when building it and editing the build files

**Comment by KurtzPT** [comment]
Usually unraid has a docker template (it's a xml file) where you define the env variables to pass to the docker container.

So lets say for example the volumes on docker-compose file, these are hardcoded and won't work on unraid. Usually these volume paths are defined directly in the docker template via unraid webpage and the docker container will use whatever values you define for each volume there.

Most likely my explanation sucks so I'll point you some links where you can figure out better what I'm trying to say.

Example of a docker template for unraid (Deluge docker): https://github.com/binhex/docker-templates/blob/master/binhex/deluge.xml

The deluge docker container that uses the above template: https://github.com/binhex/arch-deluge

As you can see the above docker container I can define the values for the volumes, PUID, PGID and even UMASK. everything is defined via the docker template (for unraid) so I don't have to build anything after I get TA image from dockerhub for example.

**Comment by MrBrax** [comment]
I can't really find out how to do that, answers online say that it's only possible at build time, especially since I'm extending an existing nginx container

**Comment by KurtzPT** [comment]
My knowledge of docker is very limited so I can't really help out much more than this.

The above docker container I gave as an example it seems to apply this stuff here at build:

https://github.com/binhex/arch-deluge/blob/master/build/root/install.sh

**Comment by MrBrax** [comment]
Seems very hacky, not sure I could really implement all that

**Comment by KurtzPT** [comment]
It's not hacky, that's the norm for docker containers, every docker container will allow you to set env variables so you can define where you want to map the volumes and other configurations that might be necessary for runtime.

The way TA is setup right now doesn't allow you to override the volumes (not even talking about uid/gid), unless you edit the docker-compose file. And if you get the image directly from dockerhub then you can't even edit that at all.

**Comment by MrBrax** [comment]
searching and replacing files in an install script looks pretty hacky to me

**Comment by KurtzPT** [comment]
I would like to help but can't help more with it unfortunately. 

The way I have it running at the moment is way more hacky then allowing users to set volumes/uid/gid at build times.

I would love to hear out other peoples TA setups to know how they tackle these issues.

**Comment by MrBrax** [comment]
yeah i don't have enough experience with docker to make something like this i think, making docker work like you want it feels like an endless battle with monkey patching and workarounds

**Comment by MrBrax** [comment]
i pushed an update that sets files to configurable permissions, but no idea if it works or if it would break the web server from reading them

**Comment by KurtzPT** [comment]
I'll try to test it out later today

**Comment by Serph91P** [comment]
Have the same problem, tried to transfer the docket setup to unRAID XML but
failed. ATM running docker compose and accessing the docket directly if I
need to do something with the files. Would welcome the change if the UID
etc. are set. Like deleting a file is already a task of a min, instead of a
couple of sec. Would love to help as well but my knowledge of Dockers and
so one is limited so far.

David ***@***.***> schrieb am Mo., 7. Feb. 2022, 13:31:

**Comment by KurtzPT** [comment]
@MrBrax I believe that I can set the chown group as an ID instead of the name of the group right? I'm trying out now, waiting for someone to start streaming.

**Comment by MrBrax** [comment]
i think so yeah

**Comment by KurtzPT** [comment]
What kind of default chmod is that "509" ?

**Comment by MrBrax** [comment]
![1644242789566](https://user-images.githubusercontent.com/1517911/152803195-4920cc2c-7a67-4084-a084-4c6eb2e2f160.png)
maybe some weird interpretation of a zero prepending an int

**Comment by KurtzPT** [comment]
Oh weird, it was showing on as 509:

![image](https://user-images.githubusercontent.com/14930057/152803474-f14fbde1-e108-4677-9f14-f0266f6826de.png)

**Comment by KurtzPT** [comment]
It's not working:

NOTICE: PHP message: PHP Warning: chown(): Unable to find uid for nobody:100 in /var/www/twitchautomator/src/TwitchVOD.php on line 419

chmod (should have been 664):
![image](https://user-images.githubusercontent.com/14930057/152850668-fe3810ab-d291-4cc9-9c03-057297153308.png)

**Comment by MrBrax** [comment]
i assume the user/group has to exist inside the container

**Comment by KurtzPT** [comment]
> i assume the user/group has to exist inside the container

I didn't change UID that's the default user.

**Comment by MrBrax** [comment]
yeah but 100

**Comment by KurtzPT** [comment]
> yeah but 100

That's the GID, the error is referring to the UID.

**Comment by MrBrax** [comment]
then i don't know, maybe some quirk of docker/alpine?

**Comment by MrBrax** [comment]
oh wait there's chown and chgrp in php

**Comment by KurtzPT** [comment]
It works fine via shell inside the container:

![image](https://user-images.githubusercontent.com/14930057/152851717-276f3e43-c2d9-42ed-a4de-895eac9b450e.png)

**Comment by MrBrax** [comment]
i split it up now

**Comment by KurtzPT** [comment]
Going to try.

**Comment by KurtzPT** [comment]
`
Type: TypeError
Code: 0
Message: octdec(): Argument #1 ($octal_string) must be of type string, int given
File: /var/www/twitchautomator/src/TwitchVOD.php

6
Trace: #0 /var/www/twitchautomator/src/TwitchVOD.php(416): octdec()
#1 /var/www/twitchautomator/src/TwitchVOD.php(1241): App\TwitchVOD->setPermissions()
#2 /var/www/twitchautomator/src/TwitchAutomator.php(735): App\TwitchVOD->saveJSON()
#3 /var/www/twitchautomator/src/TwitchAutomator.php(432): App\TwitchAutomator->download()
#4 /var/www/twitchautomator/src/Controller/ApiController.php(736): App\TwitchAutomator->handle()

#5 /var/www/twitchautomator/vendor/slim/slim/Slim/Handlers/Strategies/RequestResponse.php(43): App\Controller\ApiController->hook()

`

**Comment by MrBrax** [comment]
the wonders of no ide type hinting

**Comment by KurtzPT** [comment]
> NOTICE: PHP message: PHP Warning: chgrp(): Unable to find gid for 100 in /var/www/twitchautomator/src/TwitchVOD.php on line 422

I mean via shell I can set with GID 100... if php can't change gids for groups outside of the container it's obsolete, that's the whole point.

**Comment by KurtzPT** [comment]
I guess I could set chmod to 777 which is bonkers but at least I would be able to remove/edit files as a guest. But chmod isn't applied to the .TS file.

**Comment by MrBrax** [comment]
it might be differentiating strings and ints, not something very common for php but who knows

**Comment by MrBrax** [comment]
alright last push probably, i'm not sure working around this

>  Attempts to change the owner of the file filename to user user. Only the superuser may change the owner of a file. 

is a good idea

**Comment by KurtzPT** [comment]
Ok it seems to be working now, except the .ts file which isn't either chmod or chown.:

![image](https://user-images.githubusercontent.com/14930057/152861063-e3260a6c-83db-43a9-a0d6-42a2187cb0dc.png)

**Comment by MrBrax** [comment]
was the permissions reset before that happened? it's not set on every page load

**Comment by KurtzPT** [comment]
I didn't reset anything, I just updated to last commit, ran the docker, started a manual recording and checked after the files.

**Comment by MrBrax** [comment]
the ts file won't be changed yeah, only after it's finalized with all the bundled files

**Comment by KurtzPT** [comment]
Alright, I mean its a pretty hacky way of dealing with permissions partially, but I guess it works. 

I'll probably take some time when I can to learn how to properly add the uid/gid on the docker itself and issue a pull request if I'm successfully

**Comment by MrBrax** [comment]
good luck!
